src/Compiler/Optimize/LowerComputedCollections.fs

```diff
@@ -252,15 +252,77 @@ let (|SeqToArray|_|) g expr =
     if g.ListCollector_tcr.CanDeref then
+        let constListSizeThreshold = 100
+        let constArraySizeThreshold = int System.UInt16.MaxValue
 
         match overallExpr with
+        // [5..1] → []
+        | SeqToList g (OptionalCoerce (OptionalSeq g amap (ConstInt32Range g (start, finish))), m) when
+            start > finish
+            ->
+            Some (mkUnionCaseExpr (g.nil_ucref, [g.int32_ty], [], m))
+
+        // [1..5] → [1; 2; 3; 4; 5] ≡ 1 :: 2 :: 3 :: 4 :: 5 :: []
+        | SeqToList g (OptionalCoerce (OptionalSeq g amap (ConstInt32Range g (start, finish))), _) when
+            finish - start < constListSizeThreshold
+            ->
+            let rec conses acc n =
+                if n < start then acc
+                else conses (mkCons g g.int32_ty (Expr.Const (Const.Int32 n, Text.Range.range0, g.int32_ty)) acc) (n - 1)
+
+            Some (conses (mkNil g Text.Range.range0 g.int32_ty) finish)
+
+        // [start..finish] → List.init (finish - start + 1) ((+) start)
+        | SeqToList g (OptionalCoerce (OptionalSeq g amap (Int32Range g (start, finish))), m) ->
+            let diff = mkAsmExpr ([AbstractIL.IL.AI_sub], [], [finish; start], [g.int32_ty], Text.Range.range0)
+            let range = mkAsmExpr ([AbstractIL.IL.AI_add], [], [diff; mkOne g Text.Range.range0], [g.int32_ty], Text.Range.range0)
+            let v, e = mkCompGenLocal Text.Range.range0 "i" g.int32_ty
+            let body = mkAsmExpr ([AbstractIL.IL.AI_add], [], [start; e], [g.int32_ty], Text.Range.range0)
+            let initializer = mkLambda Text.Range.range0 v (body, g.int32_ty)
+            let init = mkCallListInit g m g.int32_ty range initializer
+            Some init
+
         | SeqToList g (OptionalCoerce (OptionalSeq g amap (overallSeqExpr, overallElemTy)), m) ->
             let collectorTy = g.mk_ListCollector_ty overallElemTy
             LowerComputedListOrArraySeqExpr tcVal g amap m collectorTy overallSeqExpr
         
+        // [|5..1|] → [||]
+        | SeqToArray g (OptionalCoerce (OptionalSeq g amap (ConstInt32Range g (start, finish))), m) when
+            start > finish
+            ->
+            Some (mkArray (g.int32_ty, [], m))
+
+        // [|1..5|] → [|1; 2; 3; 4; 5|]
+        | SeqToArray g (OptionalCoerce (OptionalSeq g amap (ConstInt32Range g (start, finish))), m) when
+            finish - start < constArraySizeThreshold
+            ->
+            Some (mkArray (g.int32_ty, [for n in start..finish -> Expr.Const(Const.Int32 n, Text.Range.range0, g.int32_ty)], m))
+
+        // [|start..finish|] → Array.init (finish - start + 1) ((+) start)
+        | SeqToArray g (OptionalCoerce (OptionalSeq g amap (Int32Range g (start, finish))), m) ->
+            let diff = mkAsmExpr ([AbstractIL.IL.AI_sub], [], [finish; start], [g.int32_ty], Text.Range.range0)
+            let range = mkAsmExpr ([AbstractIL.IL.AI_add], [], [diff; mkOne g Text.Range.range0], [g.int32_ty], Text.Range.range0)
```

Comment:
I guess this really needs to emit the equivalent of `max (finish - start + 1) 0`, since we can't know at compile time that `start` will be less than or equal to `finish`.

Reply:
0a23a5e, although there are some extra `ldnull`s in there?